File: packages/core-php/src/ES256.php
<?php

declare(strict_types=1);

namespace OpenPasskey;

class ES256
{
    public static function verify(string $coseKeyData, string $authData, string $clientDataJSON, string $signature): void
    {
        $key = self::decodeCoseKey($coseKeyData);
        self::verifyWithKey($key, $authData, $clientDataJSON, $signature);
    }

    public static function verifyWithKey(string $pemKey, string $authData, string $clientDataJSON, string $signature): void
    {
        $clientDataHash = hash('sha256', $clientDataJSON, true);
        $verifyData = $authData . $clientDataHash;

        $result = openssl_verify($verifyData, $signature, $pemKey, OPENSSL_ALGO_SHA256);

        if ($result !== 1) {
            throw new WebAuthnException('signature_invalid');
        }
    }

    public static function verifyWithRawHash(string $pemKey, string $verifyData, string $signature): void
    {
        $result = openssl_verify($verifyData, $signature, $pemKey, OPENSSL_ALGO_SHA256);
        if ($result !== 1) {
            throw new WebAuthnException('signature_invalid');
        }
    }

    public static function decodeCoseKey(string $data): string
    {
        $map = CborDecoder::decode($data);

        $kty = $map[1] ?? null;
        $alg = $map[3] ?? null;
        $crv = $map[-1] ?? null;
        $x = $map[-2] ?? null;
        $y = $map[-3] ?? null;

        if ($kty !== Cose::KTY_EC2 || $alg !== Cose::ALG_ES256 || $crv !== 1) {
            throw new WebAuthnException('unsupported_cose_algorithm');
        }
        if (strlen($x) !== 32 || strlen($y) !== 32) {
            throw new WebAuthnException('unsupported_cose_algorithm', 'Invalid EC2 key coordinate length');
        }

        return self::rawToPem($x, $y);
    }

    public static function rawToPem(string $x, string $y): string
    {
        $uncompressed = "\x04" . $x . $y;

        // ASN.1 DER encoding for EC P-256 public key
        $ecParams = hex2bin('06082a8648ce3d030107'); // OID 1.2.840.10045.3.1.7 (P-256)
        $algId = hex2bin('06072a8648ce3d0201'); // OID 1.2.840.10045.2.1 (ecPublicKey)
        $bitString = "\x03" . chr(strlen($uncompressed) + 1) . "\x00" . $uncompressed;
        $algSeq = "\x30" . chr(strlen($algId) + strlen($ecParams)) . $algId . $ecParams;
        $outer = "\x30" . self::derLength(strlen($algSeq) + strlen($bitString)) . $algSeq . $bitString;

        return "-----BEGIN PUBLIC KEY-----\n" . chunk_split(base64_encode($outer), 64, "\n") . "-----END PUBLIC KEY-----\n";
    }

    private static function derLength(int $len): string
    {
        if ($len < 128) {
            return chr($len);
        }
        $bytes = '';
        $temp = $len;
        while ($temp > 0) {
            $bytes = chr($temp & 0xFF) . $bytes;
            $temp >>= 8;
        }
        return chr(0x80 | strlen($bytes)) . $bytes;
    }
}

File: packages/core-ruby/lib/open_passkey/es256.rb
# frozen_string_literal: true

require "openssl"
require "digest"

module OpenPasskey
  module ES256
    class << self
      def verify(cose_key_data, auth_data, client_data_json, signature)
        pem_key = decode_cose_key(cose_key_data)
        verify_with_key(pem_key, auth_data, client_data_json, signature)
      end

      def verify_with_key(pkey, auth_data, client_data_json, signature)
        client_data_hash = OpenSSL::Digest::SHA256.digest(client_data_json)
        verify_data = auth_data + client_data_hash
        verify_with_raw_hash(pkey, verify_data, signature)
      end

      def verify_with_raw_hash(pkey, verify_data, signature)
        result = pkey.verify("SHA256", signature, verify_data)
        raise WebAuthnError.new("signature_invalid") unless result
      rescue OpenSSL::PKey::PKeyError
        raise WebAuthnError.new("signature_invalid")
      end

      def decode_cose_key(data)
        map = CborDecoder.decode(data)

        kty = map[1]
        alg = map[3]
        crv = map[-1]
        x = map[-2]
        y = map[-3]

        unless kty == Cose::KTY_EC2 && alg == Cose::ALG_ES256 && crv == 1
          raise WebAuthnError.new("unsupported_cose_algorithm")
        end
        unless x.is_a?(String) && y.is_a?(String) && x.bytesize == 32 && y.bytesize == 32
          raise WebAuthnError.new("unsupported_cose_algorithm", "Invalid EC2 key coordinate length")
        end

        raw_to_pkey(x, y)
      end

      def raw_to_pkey(x, y)
        uncompressed = "\x04".b + x.b + y.b

        group = OpenSSL::PKey::EC::Group.new("prime256v1")
        point = OpenSSL::PKey::EC::Point.new(group, OpenSSL::BN.new(uncompressed, 2))

        asn1 = OpenSSL::ASN1::Sequence.new([
          OpenSSL::ASN1::Sequence.new([
            OpenSSL::ASN1::ObjectId.new("id-ecPublicKey"),
            OpenSSL::ASN1::ObjectId.new("prime256v1"),
          ]),
          OpenSSL::ASN1::BitString.new(point.to_octet_string(:uncompressed)),
        ])

        OpenSSL::PKey::EC.new(asn1.to_der)
      end
    end
  end
end
